File: InterviewPrep/ClimbingTheLeaderboard.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Result
{
    public static List<int> ClimbingLeaderboard(List<int> ranked, List<int> player)
    {
        // Firstly, since the rankings do not skip numbers and just give the same
        // one to all those who tie, then we can make our life easier by working
        // with a simplified ranked list with only the unique values. The exercise
        // specifies that the existing rankings are already sorted in natural or
        // descending order.

        List<int> uniqRanks = ranked.Distinct().ToList();
        List<int> newRanks = new List<int>();

        // Next, we will do a small variant of Binary Search to find the index
        // where this score would go in the rankings.

        foreach (int score in player)
        {
            if (uniqRanks.Count == 0)
            {
                newRanks.Add(1);
                continue;
            }

            int scoreRank = BinarySearchIndex(uniqRanks, score, 0, uniqRanks.Count - 1);
            newRanks.Add(scoreRank);
        }

        return newRanks;
    }

    private static int BinarySearchIndex(
        List<int> rankedScores,
        int playerScore,
        int chunkStart,
        int chunkEnd)
    {
        int chunkSize = chunkEnd - chunkStart + 1;

        if (chunkSize == 1)
        {
            return playerScore >= rankedScores[chunkStart]
                ? chunkStart + 1
                : chunkStart + 2;
        }

        // This thing insists on behaving like an edge-case, so let's treat it
        // separately as such.

        if (chunkSize == 2)
        {
            int first = rankedScores[chunkStart];
            int second = rankedScores[chunkEnd];

            if (playerScore >= first)
            {
                return chunkStart + 1;
            }
            else if (playerScore < first && playerScore >= second)
            {
                return chunkStart + 2;
            }
            else
            {
                return chunkStart + 3;
            }
        }

        int chunkMid = chunkStart + (chunkSize / 2);
        int testScore = rankedScores[chunkMid];

        if (playerScore == testScore)
        {
            return chunkMid + 1;
        }

        if (playerScore > testScore)
        {
            return BinarySearchIndex(rankedScores, playerScore, chunkStart, chunkMid);
        }

        return BinarySearchIndex(rankedScores, playerScore, chunkMid, chunkEnd);
    }
}

class Solution
{
    static int Main(string[] args)
    {
        int rankedCount = Convert.ToInt32(Console.ReadLine().Trim());

        List<int> ranked = Console.ReadLine()
                                  .TrimEnd()
                                  .Split(' ')
                                  .ToList()
                                  .Select(rankedTemp => Convert.ToInt32(rankedTemp))
                                  .ToList();

        int playerCount = Convert.ToInt32(Console.ReadLine().Trim());

        List<int> player = Console.ReadLine()
                                  .TrimEnd()
                                  .Split(' ')
                                  .ToList()
                                  .Select(playerTemp => Convert.ToInt32(playerTemp))
                                  .ToList();

        List<int> result = Result.ClimbingLeaderboard(ranked, player);
        Console.WriteLine(String.Join("\n", result));

        return 0;
    }
}

File: CompetitiveProgramming/Hackerrank/Medium/climbing_the_leaderboard.py
#!/usr/bin/env python3

def climbing_leaderboard(ranked: list[int], player: list[int]) -> list[int]:
    # Since equal scores receive the exact same ranking, then we will remove the
    # duplicates for easier processing of the list.

    seen = set()
    ranked_uniq = [r for r in ranked if r not in seen and (seen.add(r) or True)]
    placements = []

    # Since we know the rankings are in natural or descending order, we can apply
    # a small variant of Binary Search to find the index where the next player
    # score would go in the rankings.

    for score in player:
        # If there are no rankings, then no matter what the score is, the next
        # player ranking will be 1st place.
        if not ranked_uniq:
            placements.append(1)
            continue

        score_rank = binary_search_index(ranked_uniq, score, 0, len(ranked_uniq) - 1)
        placements.append(score_rank)

    return placements


def binary_search_index(ranked_scores: list[int],
                        player_score: int,
                        chunk_start: int,
                        chunk_end: int) -> int:
    chunk_size = chunk_end - chunk_start + 1

    # If the chunk size is 1, then the answer is its index if the player score is
    # greater than or equal to the chunk's value, or the next position if it is
    # smaller than it. Well, we use +1 and +2 to compensate for the 0-based indexing
    # of lists in Python.

    if chunk_size == 1:
        return chunk_start + 1 if player_score >= ranked_scores[chunk_start] \
            else chunk_start + 2

    # This thing insists on behaving like an edge-case, so let's treat it
    # separately as such.

    if chunk_size == 2:
        first = ranked_scores[chunk_start]
        second = ranked_scores[chunk_end]

        if player_score >= first:
            return chunk_start + 1

        if first > player_score >= second:
            return chunk_start + 2

        return chunk_start + 3

    # Binary Search Recurse!

    chunk_mid = chunk_start + int(chunk_size / 2)
    test_score = ranked_scores[chunk_mid]

    if player_score == test_score:
        return chunk_mid + 1

    if player_score > test_score:
        return binary_search_index(ranked_scores, player_score, chunk_start, chunk_mid)

    return binary_search_index(ranked_scores, player_score, chunk_mid, chunk_end)


def main():
    ranked_count = int(input().strip())
    ranked = list(map(int, input().rstrip().split()))

    player_count = int(input().strip())
    player = list(map(int, input().rstrip().split()))

    result = climbing_leaderboard(ranked, player)
    print('\n'.join(map(str, result)))


if __name__ == '__main__':
    main()
